Mathlib/Topology/Algebra/Valued/WithZeroMulInt.lean

```diff
@@ -45,4 +44,54 @@ lemma exists_pow_lt_of_le_exp_neg_one [Valued R ℤᵐ⁰] {x : R} (hx : v x ≤
   rw [← exp_zero, exp_lt_exp]
   norm_num
 
+variable {K : Type*} [Field K] [Valued K ℤᵐ⁰]
+
+theorem irreducible_valuation_lt_one {ϖ : 𝒪[K]} (h : Irreducible ϖ) : v ϖ.1 < 1 :=
+  lt_of_le_of_ne (Valuation.mem_integer_iff _ _ |>.1 ϖ.2) <|
+    mt (Valuation.integer.integers _).isUnit_iff_valuation_eq_one.2 h.not_isUnit
+
+theorem irreducible_valuation_le_exp_neg_one {ϖ : 𝒪[K]} (h : Irreducible ϖ) :
+    v ϖ.1 ≤ exp (-1 : ℤ) := by
+  have hϖ : v ϖ.1 ≠ 0 := by simp [h.ne_zero]
+  have := log_one (M := ℤ) ▸  (log_lt_log hϖ one_ne_zero).2 (irreducible_valuation_lt_one h)
+  rw [← log_le_iff_le_exp hϖ]
+  linarith
+
+theorem mem_maximalIdeal_pow_valuation [IsDiscreteValuationRing 𝒪[K]]
+    {x : 𝒪[K]} {n : ℕ} (hx : x ∈ 𝓂[K] ^ n) {ϖ : 𝒪[K]} (h : Irreducible ϖ) :
+    v x.val ≤ v ϖ.1 ^ n := by
+  rcases eq_or_ne x 0 with (rfl | hx₀) <;> try simp
+  simp_rw [h.maximalIdeal_eq, Ideal.span_singleton_pow, Ideal.mem_span_singleton] at hx
+  obtain ⟨y, rfl⟩ := hx
+  simpa using le_trans (mul_le_of_le_one_right' <| (Valuation.mem_integer_iff _ _).1 y.2) le_rfl
+
```

Comment:
Ideally, armed with these lemmas, you could prove that `ValuativeRel.IsDiscrete` iff `IsDiscreteValuationRing` of the valuation ring. Which would hold for Zm0 and any isomorphic type. This way the statements here would be more general and yet still hold from Zm0 out of the box.